dev/run-tests.py

```diff
@@ -571,8 +572,8 @@ def main():
     # backwards compatibility checks
     if build_tool == "sbt":
         # Note: compatibility tests only supported in sbt for now
-        detect_binary_inop_with_mima()
-        # Since we did not build assembly/assembly before running dev/mima, we need to
+        detect_binary_inop_with_mima(hadoop_version)
+        # Since we did not build assembly/package before running dev/mima, we need to
         # do it here because the tests still rely on it; see SPARK-13294 for details.
         build_spark_assembly_sbt(hadoop_version)
```

Comment:
I'm not sure about how these two really interact; I noticed that `detect_binary_inop_with_mima` would, directly or indirectly, end up running the `package` target for the build. Also note there are still some assemblies that are built, for streaming connectors.

Overall, because of SPARK-9284, we could potentially change all this to run everything in one step (build, tests and potentially even mima checks), although that could potentially make error messages less clear. I'd rather look at that separately though.